sdk/identity/azure-identity/src/main/java/com/azure/identity/IdentityClient.java

```diff
@@ -123,34 +145,143 @@ public IdentityClient(IdentityClientOptions options) {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        try {
+            PublicClientApplication application = PublicClientApplication.builder(clientId)
+                .authority(authorityUrl)
+                .executorService(executor)
+                .build();
+            return Mono.fromFuture(application.acquireToken(UserNamePasswordParameters.builder(new HashSet<>(Arrays.asList(scopes)), username, password.toCharArray()).build()))
+                .map(ar -> new AccessToken(ar.accessToken(), OffsetDateTime.ofInstant(ar.expiresOnDate().toInstant(), ZoneOffset.UTC)))
+                .doFinally(s -> executor.shutdown());
+        } catch (MalformedURLException e) {
+            return Mono.error(e);
+        }
+    }
+
+    /**
+     * Asynchronously acquire a token from Active Directory with a device code challenge. Active Directory will provide
+     * a device code for login and the user must meet the challenge by authenticating in a browser on the current or a
+     * different device.
+     *
+     * @param tenantId           the tenant ID of the application
+     * @param clientId           the client ID of the application
+     * @param scopes             the scopes to authenticate to
+     * @param deviceCodeConsumer the user provided closure that will consume the device code challenge
+     * @return a Publisher that emits an AccessToken when the device challenge is met, or an exception if the device code expires
+     */
+    public Mono<AccessToken> authenticateWithDeviceCode(String tenantId, String clientId, String[] scopes, Consumer<DeviceCodeChallenge> deviceCodeConsumer) {
+        String authorityUrl = options.authorityHost().replaceAll("/+$", "") + "/" + tenantId;
+        ExecutorService executor = Executors.newSingleThreadExecutor();
         try {
-            context = new AuthenticationContext(authorityUrl, false, executor);
-        } catch (MalformedURLException mue) {
-            throw Exceptions.propagate(mue);
+            PublicClientApplication.Builder applicationBuilder = PublicClientApplication.builder(clientId)
+                .authority(authorityUrl)
+                .executorService(executor);
+            if (options.proxyOptions() != null) {
+                applicationBuilder.proxy(proxyOptionsToJavaNetProxy(options.proxyOptions()));
+            }
+            PublicClientApplication application = applicationBuilder.build();
+            return Mono.fromFuture(() -> {
+                DeviceCodeFlowParameters parameters = DeviceCodeFlowParameters.builder(new HashSet<>(Arrays.asList(scopes)), deviceCodeInternal -> deviceCodeConsumer.accept(new DeviceCodeChallenge(deviceCodeInternal))).build();
+                return application.acquireToken(parameters);
+            })
+                .map(ar -> new AccessToken(ar.accessToken(), OffsetDateTime.ofInstant(ar.expiresOnDate().toInstant(), ZoneOffset.UTC)))
```

Comment:
Put the `.map` operation at the end of the previous line, rather than on a new line.